Allow multipe DNS servers as comma-seperated argument for kubelet --dns

rtreffer opened the pull request:
This PR explores how kubectls "--dns" could be extended to specify multiple DNS servers for in-cluster PODs. Testing on the local libvirt-coreos cluster shows that multiple DNS server are injected without issues.

Specifying multiple DNS servers increases resilience against
- Packet drops
- Single server failure

I am debugging services that do 50+ DNS requests for a single incoming interactive request, thus highly increase the chance of a slowdown (+5s) due to a single packet drop. Switching to two DNS servers will reduce the impact of the issues (roughly +1s on glibc, 0s on musl, error-rate goes down to error-rate^2).

Note that there is no need to change any runtime related code as far as I know. In the case of "default" dns the /etc/resolv.conf is parsed and multiple DNS server are send to the backend anyway. This only adds the same capability for the clusterFirst case.

I've heard from @thockin that multiple DNS entries are somehow considered. I've no idea what was considered, though. This is what I would like to see for our production use, though.

```release-note
NONE
```

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/33550)

<!-- Reviewable:end -->

rtreffer commented:
There is a signed company CLA and I'm part of the correct group as far as I know, any reason for the CLA to fail?

mkumatag commented on pkg/kubelet/kubelet.go:
```diff
@@ -435,6 +435,15 @@ func NewMainKubelet(kubeCfg *componentconfig.KubeletConfiguration, kubeDeps *Kub
 		myFlannelExperimentalOverlay = false
 	}
 
+	clusterDNSIps := strings.Split(kubeCfg.ClusterDNS, ",")
```
IMO instead of having own logic of splitting the string, we can user StringSliceVar function @ https://github.com/kubernetes/kubernetes/blob/master/cmd/kubelet/app/options/options.go#L138

vishh commented on pkg/apis/componentconfig/types.go:
```diff
@@ -224,10 +224,10 @@ type KubeletConfiguration struct {
-	// addition to the host's DNS servers
-	ClusterDNS string `json:"clusterDNS"`
+	// clusterDNS is a list of IP address for a cluster DNS server.  If set,
+        // kubelet will configure all containers to use this for DNS resolution
```
nit: Fix format

vishh reviewed (changes requested):
Fix the nits, otherwise LGTM. cc @kubernetes/sig-network

bowei commented on pkg/kubelet/kubelet.go:
```diff
@@ -396,6 +396,14 @@ func NewMainKubelet(kubeCfg *componentconfig.KubeletConfiguration, kubeDeps *Kub
+		ip := net.ParseIP(ipstring)
+		if ip != nil {
+			clusterDNS = append(clusterDNS, ip)
+		}
```
````
if ip == nil {
  glog.Warningf("Invalid clusterDNS ip '%q'", ipstring)
} else {
  ...
}
````
if the ip does not parse

rtreffer replied:
I actually spend a few hours debugging a broken ClusterDNS IP (IP + space instead of IP)....
PR is updated. Good point.

bowei commented on cmd/kubelet/app/options/options.go:
```diff
@@ -135,7 +135,7 @@ func (s *KubeletServer) AddFlags(fs *pflag.FlagSet) {
 	fs.StringVar(&s.ClusterDomain, "cluster-domain", s.ClusterDomain, "Domain for this cluster.  If set, kubelet will configure all containers to search this domain in addition to the host's search domains")
 	fs.StringVar(&s.MasterServiceNamespace, "master-service-namespace", s.MasterServiceNamespace, "The namespace from which the kubernetes master services should be injected into pods")
-	fs.StringVar(&s.ClusterDNS, "cluster-dns", s.ClusterDNS, "IP address for a cluster DNS server.  This value is used for containers' DNS server in case of Pods with \"dnsPolicy=ClusterFirst\"")
+	fs.StringSliceVar(&s.ClusterDNS, "cluster-dns", s.ClusterDNS, "Comma-seperated list of DNS server IP address.  This value is used for containers DNS server in case of Pods with \"dnsPolicy=ClusterFirst\"")
```
separated (spelling)

thockin commented:
Is your intention to run two DNS Services in the cluster and use those
here?  will they use the same DNS backends or run different sets of pods?

I know musl likes to ask every DNS server in parallel (doubling the DNS
load on your cluster), but others do not.  So, how does this actually make
things better?  If a packet is lost or a backend is dead (or slow) the
client has to time out, right?  Are you saying glibc does a shorter timeout
if it has a second DNS server listed?

On Sat, Oct 8, 2016 at 6:08 AM, Rene Treffer <notifications@github.com>
wrote:

> *@rtreffer* commented on this pull request.
> ------------------------------
>
> In pkg/kubelet/kubelet.go
> <https://github.com/kubernetes/kubernetes/pull/33550>:
>
> > @@ -396,6 +396,14 @@ func NewMainKubelet(kubeCfg *componentconfig.KubeletConfiguration, kubeDeps *Kub
>  		myConfigureCBR0 = false
>  	}
>
> +	clusterDNS := make([]net.IP, 0, len(kubeCfg.ClusterDNS))
> +	for _, ipstring := range kubeCfg.ClusterDNS {
> +		ip := net.ParseIP(ipstring)
> +		if ip != nil {
> +			clusterDNS = append(clusterDNS, ip)
> +		}
>
> I actually spend a few hours debugging a broken ClusterDNS IP (IP + space
> instead of IP)....
> PR is updated. Good point.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/33550>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVE7Ro7jcfBDqixdsDNFdCAGsw4YHks5qx5WxgaJpZM4KHigd>
> .
>

rtreffer commented:
@thockin retested, nslookup overwrote the timeout but it works as expected with plain ping. 
So yes, this would mainly guard against different errors and not heavily reduce latency on glibc. We have seen a few rare double/triple-drops (+10/+15s latency) which might be related to glibc not changing the source port on retry with the same server.

So the key here would be to reduce errors by having 2 server that have different error cases
- One host-local server to avoiding network drops
- One balanced service IP avoiding single service failures

I would like to also lower the default "options timeout:5" to 1 second to allow faster retries / faster switching between servers but that's a different story.

thockin commented:
What I think would work more consistently is to have a single nameserver,
the local node, which implemented timeouts, suffix-matches for upstream
nameservers, and retries in a consistent way that didn't change with
different resolver libs.

On Tue, Oct 11, 2016 at 2:10 AM, Rene Treffer <notifications@github.com>
wrote:

> @thockin <https://github.com/thockin> retested, nslookup overwrote the
> timeout but it works as expected with plain ping.
> So yes, this would mainly guard against different errors and not heavily
> reduce latency on glibc. We have seen a few rare double/triple-drops
> (+10/+15s latency) which might be related to glibc not changing the source
> port on retry with the same server.
>
> So the key here would be to reduce errors by having 2 server that have
> different error cases
>
>    - One host-local server to avoiding network drops
>    - One balanced service IP avoiding single service failures
>
> I would like to also lower the default "options timeout:5" to 1 second to
> allow faster retries / faster switching between servers but that's a
> different story.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/33550#issuecomment-252855228>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVFbQHnG6nyz3m-DK63KznUSmOkcpks5qy1KigaJpZM4KHigd>
> .
>

rtreffer commented:
I totally agree that a per-node resolver is the first thing to do. We are doing that right now with a daemonset.
This patch is for a point in the future when we have a local DNS server and would like to reduce the impact of temporary single-server issues.

What I'm trying to do would look like
```
+-----------------+                 +-----------------+
|      Node A     |  Service backed |      Node B     |
|                 |   by Daemonset  |                 |
|  Daemonset dns <-------+   +-------> Daemonset DNS  |
|    ^            |      |   |      |             ^   |
|    | DNS primary|      |   |      |             |   |
|    + ns / Query |      |   |      |             |   |
|  POD p          |      |   |      |             |   |
+----+------------+      |   |      +-----------------+
     |                   |   |                    |
     |                   |   |                    |
     |                +--v---v--+                 |
     | DNS fallback   |         | DNS fallback    |
     +--------------->+ DNS SVC +-----------------+
       Query          |         | Query
                      +---------+
```
The same DNS resolving would be used on both paths. For the major resolver libs the local DNS will win (due to latency or due to being asked first) unless there is an error / issue in which case going down another route can be helpful.

rtreffer commented:
Both CLAs signed, @vishh issues addressed and freshly rebased.

thockin commented:
While this PR seems relatively harmless on its own, I don't think I understand how you intend to use it?

Do you want to run multiple kube-dns deployments and services and reference them here?

Do you want to run one kube-dns and multiple services and reference them here?

Do you want to use one kube-dns and service and just repeat the IP here (tricky)?

Do you want to reference OTHER nameservers?

It's the last option that scares me - I don't really want to encourage that because it's easy to get wrong (or to work for some resolvers and not others).

rtreffer commented:
So far we've switched to a daemonset for DNS (caches in front of kube-dns, not kube-dns itself). This has eliminated network drops on DNS lookups and made several services work on our kubernetes cluster.

Right now we're using "--cluster-dns=$(hostname -i)"

The pods in this daemonset are also be available through a service IP (let's say 10.22.0.3).

What I want to do is add this service IP besides the local ip ("--cluster-dns=$(hostname -i),10.22.0.3").

Thus each resolver would start with either the local one, a random (but verified) remote one, or both, then alternating / retrying.

I'm not sure how that map that to the options you listed. I'd guess it's closest to have one instance pool but specifying it twice as the backing pods are always the daemonset pods.

thockin commented:
Interesting approach.  I guess I am OK with it - it seems compatible...  I want @bowei to have a look.

bowei commented:
implementation looks ok, I added a warning regarding potential undefined behavior if the DNS servers return different results

rtreffer commented:
@bowei rebased & warning added.

therc commented:
I wanted to implement this and even mentioned it to @mrubin last night. Even with DNS v20, which has improved things through graceful termination, we see hiccups in applications (slowness, if not failures), due to pods restarting. My plan was to have two separate cluster services backed by separate deployments. I'd rather take a little latency hit here and there (for glibc images) than lookup failures.

bowei commented:
@bprashanth

bprashanth commented:
This missed 1.5 anyway, which hit code freeze last night. Earliest it can get in is 1.6, for which the tree opens end November (~20 I think), so I'm inclined to let it sit for now.

bowei commented:
@vishh do you want to remove the "changes requested"? This can go in after the code freeze.

rtreffer commented:
Freshly rebased, according to https://github.com/kubernetes/features/blob/master/release-1.5/release-1.5.md the 1.6 master should be opened now?

bowei commented:
/lgtm

bowei commented:
@k8s-bot verify all

rtreffer-sc committed:
Allow multipe DNS servers as comma-seperated argument for --dns

Depending on an exact cluster setup multiple dns may make sense.
Comma-seperated lists of DNS server are quite common as DNS servers
are always plain IPs.

bowei commented:
@vishh -- can you remove the "changes requested" status from this review?
@rtreffer -- files need to be merged and the changes repushed. It would be great to get this resolved.

rtreffer commented:
@bowei pushed. I reseted the generated files to clean up the diff, too.

bowei commented:
@k8s-bot verify all

bowei commented:
@k8s-bot verify all

bowei commented:
@k8s-bot verify test this

rmmh commented:
@k8s-bot test this

bowei commented:
/lgtm

bowei commented:
/approve

dchen1107 commented:
/approve
